### PR TITLE
[WIP] Tests for fuzzy_match_multi and fuzzy_match_windows

### PR DESCRIPTION
**Pull request by aarthykc**
Branched off `window-weld`, this PR will add the following: 

- lazy_static that constructs FuzzyPhraseSet, which I did using `ensure-data` in `test_utils`
- Partial_eq for `FuzzyMatchResult` and `FuzzyWindowResult` to check if the phrases un the result are equivalent
- test to check if `fuzzy_match_windows` and `fuzzy_match_multi` give the same result when a "damaged" (phrase that contains a typo) is passed.

cc @boblannon

### CODE REVIEW

**Comment by apendleton** [comment]
@aarthykc what I've done so far on the glue code is to have two buckets of tests:
* the ones in `glue/mod.rs` that run quickly and targets specific interesting test cases we know we want to be sure to hit
* the ones in `glue/fuzz_tests.rs` one that require a bunch of data and run slowly, that we can run to make sure we cover any unanticipated corner cases, but are optional so that tests still run in a reasonable amount of time (with the idea being that any corner cases we catch, we can write specific tests for in the unit-tests section). They also pretty much have to be built in `release` mode to complete in a reasonable amount of time, which is yet another reason not to run them every time. These can be run with `cargo test --release glue_fuzztest -- --ignored --nocapture --test-threads=1`

I'm not dead-set on that structure, but would ideally prefer not to impose a requirement that every test run downloads a several-hundred-meg file from s3. Ideally CI runs should be pretty quick and development in a data-constrained environment should be possible.

It seems like what's going on here is somewhere in between the two: these tests aren't necessarily focused on particular cases, but there aren't enough of them for us to be confident that we're covering our bases just by dumb luck. My inclination is probably to add some exhaustive tests of this functionality in `fuzz_tests.rs`, then (I think we could achieve this pretty quickly by borrowing code from the benchmarks I wrote today and just checking that their output is correct), and to instead make these tests a little more specific. If we do, however, choose to stick with building a test set full of dummy data in the regular tests, I think we'd be better off using some of the data that's already committed to the repo, in `benches/data/*`, rather than downloading a giant file just to use 20 lines of it.

File: src/glue/mod.rs [hunk 1]
```diff
@@ -182,6 +182,12 @@ pub struct FuzzyWindowResult {
     pub ends_in_prefix: bool,
 }
 
+impl<'a, 'b> PartialEq<FuzzyMatchResult> for FuzzyWindowResult {
+    fn eq(&self, other: &FuzzyMatchResult) -> bool {
+        self.phrase == other.phrase
+    }
+}
```
Comment (apendleton) [inline src/glue/mod.rs]: This should also compare the edit distance, since they both share that field as well.

**Comment by apendleton** [comment]
@aarthykc I also meant to ask: if you want to focus on unit tests I think I can turn what I've already written in terms of benchmarks into fuzz tests as well, or you can do both, whichever you prefer. Not sure which parts you've already started.

**Comment by aarthykc** [comment]
@apendleton added the changes you requested, I know we went through most of these yesterday but would like another review and if it seems fine, please merge 🌮

**Comment by apendleton** [comment]
@aarthykc these looked great. One concern I had:
```
//check if windowed match and multi windowed match give the same results
assert!(windowed_match_result.iter().any(|x| {
    windowed_match_multi_result.iter().any(|y| {
        y.iter().any(|z| {
            x == z
        })
    })
}));
```
This code makes a fairly weak guarantee: it's true if any result in the windowed search has a corresponding result in the multi search, but actually, *every* result should match. Further, the order matters for the multi result, because the position in the response corresponds to the `start_position` property on results that come back from windowed search, so there's more to check than just the phrase and the edit distance.

I went ahead and fixed this, just in the interests of getting this merged and avoiding taking up another day of your time on back and forth, especially because the significance of the order, etc., is not well-documented and I think this would have been hard to figure out.

I also copy-pasted this function to make a version that does prefix search rather than just full search... and once I did that, I discovered a bug in the prefix-handling code of windowed search 🤦‍♂️ (my code, not yours). So I fixed that too.
